nixos/modules/programs/neovim.nix

```diff
@@ -0,0 +1,26 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.programs.neovim;
+
+in {
+  meta.maintainers = [ maintainers.matoruru ];
+
+  options.programs.neovim = {
+    defaultEditor = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        When enabled, installs neovim(nvim) and configures neovim(nvim) to be the default editor
+        using the EDITOR environment variable.
+      '';
+    };
+  };
+
+  config = mkIf cfg.defaultEditor {
+    environment.systemPackages = [ pkgs.neovim ];
+    environment.variables = { EDITOR = mkOverride 900 "nvim"; };
```

Comment:
I personally don't think it's worth having an option just for this. However if this module gained a way to configure neovim as described [in the manual](https://nixos.org/nixpkgs/manual/#users-guide-to-vim-pluginsaddonsbundlesscripts-in-nixpkgs), with options documenting the possible customizations, this would be much more worthwhile.

Reply:
Thank you maintainer,
I understand what you mean. But the problem I think is just that programs.neovim module does not exist.

Reply:
And now we do not have the way to make neovim the default editor on our environment, except creating custom configuration.

Reply:
You don't need a separate module for setting these options. You can just put

```nix
{ pkgs, ... }: {
  environment.systemPackages = [ pkgs.neovim ];
  environment.variables.EDITOR = "nvim";
}
```

in your `configuration.nix` with the same effect

Reply:
Thank you to show me that way. But I wonder why neovim or other editors do not have the option, defaultEditor, even though emacs and vim have it.

Reply:
Ah you're right, `vim` does have such an exact same module without any extra functionality either. But I don't like how this is done, what if I set both emacs and vim to be the default editor? It would give an error. A better way would be to have a single option like `programs.defaultEditor = pkgs.vim`, which would be a generalization I wouldn't mind having.

Reply:
That makes sense. I understood what you mean and fully agree that. I also think that is better way.

Reply:
Thank you to review this and explain the details.

Reply:
But I notice about this:

> You don't need a separate module for setting these options. You can just put
> 
> ```nix
> { pkgs, ... }: {
>   environment.systemPackages = [ pkgs.neovim ];
>   environment.variables.EDITOR = "nvim";
> }
> ```
> 
> in your `configuration.nix` with the same effect

And I tried this:

```
environment.variables.EDITOR = "aaaaaaaaaaaaaaa";
```

this has done with no error of course. I think it is **unsafe**. if we set both emacs and vim to be the default editor Nix is going to give an error as you said, but it is **safe**.

Reply:
There's nothing special about the string `nvim` that makes it safe though. What if you created an editor whose binary is called "aaaaaaaaaaaaaaaaaa"?

With an option like `programs.defaultEditor = pkgs.neovim`, you could mark the `neovim` package itself as an editor where the binary is called `nvim`, like
```nix
stdenv.mkDerivation {
  name = "neovim";
  # ...
  passthru = {
    hasEditor = "nvim";
  };
}
```

Then the module can do an assertion that the options value has the `hasEditor` attribute, and set the `EDITOR` variable to its value, like this:

```nix
{
  assertions = [{
    assertion = programs.defaultEditor ? hasEditor;
    message = "The package defined in programs.defaultEditor is not an editor";
  }];
  
  environment.systemPackages = [ programs.defaultEditor ];
  environment.EDITOR = programs.defaultEditor.hasEditor;

}
```

That would then
- Be safe, as you can't assign packages that aren't editors
- Be safe, because it would be obvious that you can't assign different values to a single option
- Not clutter the module and option listing with a new option for every editor

Reply:
Yes both "nvim" and "aaaaaaaaaaaaaaa" are not special as you said. So I thought we need some constraints to set defaultEditor safely.

Reply:
That is great. I did not have a way like that.

Reply:
> Yes both "nvim" and "aaaaaaaaaaaaaaa" are not special as you said. So I thought we need some constraints to set defaultEditor safely.

Ah right, I mixed something up there

Reply:
I see. Thank you for this discussion. I also guess this PR is not useful. It is better to go a another way like you said on a another PR.

Reply:
Could you close this PR?